src/SonarQube.Client/Services/SonarQubeClient.cs

```diff
@@ -0,0 +1,264 @@
+﻿using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Newtonsoft.Json.Linq;
+using SonarQube.Client.Helpers;
+using SonarQube.Client.Models;
+
+namespace SonarQube.Client.Services
+{
+    public class SonarQubeClient : ISonarQubeClient
+    {
+        private readonly HttpMessageHandler messageHandler;
+        private readonly TimeSpan requestTimeout;
+
+        public SonarQubeClient(HttpMessageHandler messageHandler, TimeSpan requestTimeout)
+        {
+            this.messageHandler = messageHandler;
+            this.requestTimeout = requestTimeout;
+        }
+
+        public Task<Result<ComponentDTO[]>> GetComponentsSearchProjectsAsync(ConnectionDTO connection,
+            ComponentRequest request, CancellationToken token)
+        {
+            const string SearchProjectsAPI = "api/components/search_projects"; // Since 6.2; internal
+
+            return SafeUseHttpClient(connection,
+                async client =>
+                {
+                    var query = AppendQueryString(SearchProjectsAPI, "?organization={0}&p={1}&ps={2}&asc=true"); // TODO: should handle optional params
+                    var httpResponse = await InvokeGetRequest(client, query, token);
+                    var stringResponse = await GetStringResultAsync(httpResponse, token);
+
+                    return ProcessJsonResponse<ComponentDTO[]>(stringResponse, token);
+                });
+        }
+
+        public Task<Result<OrganizationDTO[]>> GetOrganizationsAsync(ConnectionDTO connection,
+            OrganizationRequest request, CancellationToken token)
+        {
+            const string OrganizationsAPI = "api/organizations/search"; // Since 6.2; internal
+
+            return SafeUseHttpClient(connection,
+                async client =>
```

Comment:
I don't like how similar these lambdas are in most of the methods, but cannot come up with anything better :(

Reply:
Shortest form I could came up with because C# doesn't know how to automatically cast up Task<T>` to `Task<object>` nor how to create a generic lambda.